imap/jmap_calendar.c

```diff
@@ -7707,6 +7721,276 @@ static int jmap_calendarevent_parse(jmap_req_t *req)
+                mbentry->uniqueid, cdata->dav.imap_uid);
+        r = IMAP_NOTFOUND;
+        goto done;
+    }
+
+    /* Check permissions. */
+    if (!mbentry || !jmap_hasrights_mbentry(req, mbentry, JACL_READITEMS)) {
+        r = IMAP_NOTFOUND;
+        goto done;
+    }
+
+    /* Check privacy for sharees */
+    if (strcmp(req->accountid, req->userid)) {
+        if (cdata->comp_flags.privacy != CAL_PRIVACY_PUBLIC) {
+            r = cdata->comp_flags.privacy == CAL_PRIVACY_SECRET ?
+                IMAP_NOTFOUND : IMAP_PERMISSION_DENIED;
+            goto done;
+        }
+    }
+
+    /* Open mailbox for reading */
+    r = jmap_openmbox(req, mbentry->name, &mbox, 0);
+    if (r) {
+        syslog(LOG_ERR, "jmap_openmbox(req, %s) failed: %s",
+                mbentry->name, error_message(r));
+        goto done;
+    }
+
+    /* Fetch index record for the resource. */
+    struct index_record record = { };
+    r = mailbox_find_index_record(mbox, cdata->dav.imap_uid, &record);
+    if (r) {
+        syslog(LOG_ERR, "mailbox_index_record(0x%x) failed: %s",
+                cdata->dav.imap_uid, error_message(r));
+        goto done;
+    }
+    /* Load VEVENT from record. */
+    update.oldical = record_to_ical(mbox, &record, NULL);
+    if (!update.oldical) {
+        syslog(LOG_ERR, "record_to_ical failed for record %u:%s",
+                cdata->dav.imap_uid, mbentry->name);
+        r = IMAP_INTERNAL;
+        goto done;
+    }
+    jmap_closembox(req, &mbox);
+
+    /* Find participantId */
+    icalcomponent *comp = icalcomponent_get_first_real_component(update.oldical);
+    icalcomponent_kind kind = icalcomponent_isa(comp);
+    const char *part_id = NULL;
+
+    for (; comp; comp = icalcomponent_get_next_component(update.oldical, kind)) {
+        icalproperty *prop =
+            icalcomponent_get_first_property(comp, ICAL_RECURRENCEID_PROPERTY);
+
+        if (update.eid->ical_recurid) {
+            /* Is it the correct override? */
+            if (!prop || strcmp(update.eid->ical_recurid,
+                                icalproperty_get_value_as_string(prop))) {
+                continue;
+            }
+        }
+        else if (prop) {
+            /* Not the master */
+            continue;
+        }
+
+        prop = find_attendee(comp, part_email);
+        if (prop) {
+            part_id = icalproperty_get_xparam_value(prop, JMAPICAL_XPARAM_ID);
+            break;
+        }
+    }
+
+    if (!part_id) {
+        r = HTTP_NOT_FOUND;
+        goto done;
+    }
+
+    /* Create patch */
+    struct buf buf = BUF_INITIALIZER;
+    if (update.eid->ical_recurid) {
+        /* XXX  FIX ME */
+        struct icaltimetype tt = icaltime_from_string(update.eid->ical_recurid);
+        struct jmapical_datetime dt;
+
+        jmapical_datetime_from_icaltime(tt, &dt);
+        jmapical_localdatetime_as_string(&dt, &buf);
+
+        buf_insertcstr(&buf, 0, "recurrenceOverrides/");
+        buf_putc(&buf, '/');
```

Comment:
If "recurrenceOverrides" is set, then the DateTime-formatted recurrence id also must be set here.

Reply:
It is (because my test works).  The recurrence-id is written to the buffer first (because jmapical_localdatetime_as_string() resets the buffer), and then we prepend the string.